crates/bevy_text/src/pipeline.rs

```diff
@@ -1,219 +1,382 @@
+        Self(cosmic_text::FontSystem::new_with_locale_and_db(locale, db))
+    }
+}
+
+/// A wrapper around a [`cosmic_text::SwashCache`]
+struct SwashCache(cosmic_text::SwashCache);
+
+impl Default for SwashCache {
+    fn default() -> Self {
+        Self(cosmic_text::SwashCache::new())
+    }
 }
 
-/// Render information for a corresponding [`Text`] component.
+/// The `TextPipeline` is used to layout and render [`Text`](crate::Text).
 ///
-///  Contains scaled glyphs and their size. Generated via [`TextPipeline::queue_text`].
-#[derive(Component, Clone, Default, Debug, Reflect)]
-#[reflect(Component, Default)]
-pub struct TextLayoutInfo {
-    pub glyphs: Vec<PositionedGlyph>,
-    pub logical_size: Vec2,
+/// See the [crate-level documentation](crate) for more information.
+#[derive(Default, Resource)]
+pub struct TextPipeline {
+    /// Identifies a font [`ID`](cosmic_text::fontdb::ID) by its [`Font`] [`Asset`](bevy_asset::Asset).
+    map_handle_to_font_id: HashMap<AssetId<Font>, cosmic_text::fontdb::ID>,
+    /// The font system is used to retrieve fonts and their information, including glyph outlines.
+    ///
+    /// See [`cosmic_text::FontSystem`] for more information.
+    font_system: CosmicFontSystem,
+    /// The swash cache rasterizer is used to rasterize glyphs
+    ///
+    /// See [`cosmic_text::SwashCache`] for more information.
+    swash_cache: SwashCache,
 }
 
 impl TextPipeline {
-    pub fn get_or_insert_font_id(&mut self, handle: &Handle<Font>, font: &Font) -> FontId {
-        let brush = &mut self.brush;
-        *self
-            .map_font_id
-            .entry(handle.id())
-            .or_insert_with(|| brush.add_font(handle.id(), font.font.clone()))
+    /// Utilizes [`cosmic_text::Buffer`] to shape and layout text
+    ///
+    /// Negative or 0.0 font sizes will not be laid out.
+    pub fn update_buffer(
+        &mut self,
+        fonts: &Assets<Font>,
+        sections: &[TextSection],
+        linebreak_behavior: BreakLineOn,
+        bounds: Vec2,
+        scale_factor: f64,
+        buffer: &mut CosmicBuffer,
+    ) -> Result<(), TextError> {
+        let font_system = &mut self.font_system.0;
+
+        // return early if the fonts are not loaded yet
+        let mut font_size = 0.;
+        for section in sections {
+            if section.style.font_size > font_size {
+                font_size = section.style.font_size;
+            }
+            fonts
+                .get(section.style.font.id())
+                .ok_or(TextError::NoSuchFont)?;
+        }
+        let line_height = font_size * 1.2;
+        let (font_size, line_height) = (font_size as f32, line_height as f32);
+        let metrics = Metrics::new(font_size, line_height).scale(scale_factor as f32);
+
+        let spans: Vec<(&str, Attrs)> = sections
+            .iter()
+            .filter(|section| section.style.font_size > 0.0)
+            .enumerate()
```

Comment:
I think this `enumerate` should happen before the `filter`, or the section indices will be incorrect if there's a section with a zero font size.

Reply:
The idea is to "silently" remove the sections that have invalid fontsize, so they will not be rendered, If i were to put the enumeration before, the indices would be able to point to non-existing values in the array

Reply:
checking it again, I see that the index is only set in the metadata, and i do not know what the metadata should be used for really

so maybe it should contain the before filtered index, since that is correct in the textsections of bevy, or it should be the spans of cosmic, since that is what cosmic is helping us render

Reply:
Ah, oops. Thought `section_index` was being used to look up style properties in `get_attrs`. I am also not sure about the metadata situation.

Reply:
Found the comment from cosmic-texts developer, the metadata is only there for consumers to use. It is not used internally in cosmic. 

Thus it makes more sense to actually have it reference the actual bevy section they were created from! I will fix this